src/operator/tensor/elemwise_unary_op.h

```diff
@@ -31,173 +31,381 @@
 #include "../mshadow_op.h"
 #include "../elemwise_op_common.h"
 #include "../special_functions-inl.h"
-#include "./broadcast_reduce-inl.h"
-#include "./init_op.h"
+#include "../mxnet_op.h"
+#include "cast_storage-inl.h"
 
 namespace mxnet {
 namespace op {
-template<typename xpu, typename op>
-void UnaryLaunch(const nnvm::NodeAttrs& attrs,
-                 const OpContext& ctx,
-                 const std::vector<TBlob>& inputs,
-                 const std::vector<OpReqType>& req,
-                 const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mxnet_op;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
 
-  CHECK_EQ(inputs.size(), 1U);
-  CHECK_EQ(outputs.size(), 1U);
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Kernel<op, xpu>::Launch(s, outputs[0].Size(),
-      outputs[0].dptr<DType>(), inputs[0].dptr<DType>());
-  });
-}
+class OpBase {
+ public:
+  template<int Req>
+  struct MapSetToScalar {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *out, const DType value) {
+      KERNEL_ASSIGN(out[i], Req, value);
+    }
+  };
 
-template<typename GRAD_OP>
-struct unary_bwd {
-  template<typename DType>
-  MSHADOW_XINLINE static DType Map(DType a, DType b) {
-    return DType(a*GRAD_OP::Map(b));
+ protected:
+  /*! \brief Copy blob data */
+  template<typename xpu>
+  static void inline CopyBlob(mshadow::Stream<xpu> *s,
+                              const TBlob *dest_blob,
+                              const OpReqType reqi,
+                              const TBlob& src_blob) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(src_blob.type_flag_, dest_blob->type_flag_);
+    CHECK_EQ(src_blob.shape_, dest_blob->shape_);
+    MSHADOW_TYPE_SWITCH(src_blob.type_flag_, DType, {
+      // Check if the pointers are the same (in-place operation needs no copy)
+      if (src_blob.dptr<DType>() != dest_blob->dptr<DType>()) {
+        mshadow::Copy(dest_blob->FlatTo1D<xpu, DType>(s), src_blob.FlatTo1D<xpu, DType>(s), s);
+      }
+    });
   }
-};
 
-template<typename xpu, typename OP>
-void UnaryCompute(const nnvm::NodeAttrs& attrs,
-                  const OpContext& ctx,
-                  const std::vector<TBlob>& inputs,
-                  const std::vector<OpReqType>& req,
-                  const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
-    ASSIGN_DISPATCH(out, req[0], F<OP>(inputs[0].FlatTo1D<xpu, DType>(s)));
-  });
-}
+  /*! \brief Allocate geometry-related blob data for sparse tensors */
+  static void AllocateGeometry(const NDArray *dest, const NDArray* clone_from = nullptr) {
```

Comment:
Could you also add comment that `clone_from` is supposed to be sparse NDArray as well?